src/modules/content-s3/content-s3.c

```diff
@@ -0,0 +1,287 @@
+    }
+    return;
+
+error:
+    if (errstr)
+        printf("\n\n%s\n\n", errstr);
+    if (flux_respond_error (h, msg, errno, errstr) < 0)
+        flux_log_error (h, "error responding to kvs-checkpoint.put request");
+}
+
+/* Table of message handler callbacks registered below.
+ * The topic strings in the table consist of <service name>.<method>.
+ */
+static const struct flux_msg_handler_spec htab[] = {
+    { FLUX_MSGTYPE_REQUEST, "content-backing.load",    load_cb, 0 },
+    { FLUX_MSGTYPE_REQUEST, "content-backing.store",   store_cb, 0 },
+    { FLUX_MSGTYPE_REQUEST, "kvs-checkpoint.get", checkpoint_get_cb, 0 },
+    { FLUX_MSGTYPE_REQUEST, "kvs-checkpoint.put", checkpoint_put_cb, 0 },
+    FLUX_MSGHANDLER_TABLE_END,
+};
+
+/* Destroy module context.
+ */
+static void content_s3_destroy (struct content_s3 *ctx)
+{
+    if(ctx) {
+        int saved_errno = errno;
+        flux_msg_handler_delvec (ctx->handlers);
+        free (ctx->cfg);
+        free (ctx);
+        errno = saved_errno;
+    }
+
+    s3_cleanup();
+}
+
+/* Create the s3 context, initalize the connection, and
+ * create the working bucket
+ */
+static struct content_s3 *content_s3_create (flux_t *h)
+{
+    const char *errstr = NULL;
+    struct content_s3 *ctx;
+
+    if (!(ctx = calloc (1, sizeof (*ctx))))
+        return NULL;
+    ctx->h = h;
+
+    if (!(ctx->hashfun = flux_attr_get (h, "content.hash"))) {
+        flux_log_error (h, "content.hash");
+        goto error;
+    }
+    struct s3_config *cfg;
+
+    if (!(cfg = calloc (1, sizeof (*cfg))))
+        goto error;
+    
+    cfg->retries = 5;
+    cfg->bucket = getenv("S3_BUCKET");
+    cfg->access_key = getenv("S3_ACCESS_KEY_ID");
+    cfg->secret_key = getenv("S3_SECRET_ACCESS_KEY");
+    cfg->hostname = getenv("S3_HOSTNAME");
+
+    if (s3_init (cfg, &errstr) < 0) {
+        flux_log_error (h, "content-s3 init");
+        goto error;
+    }
+
+    if (s3_bucket_create (cfg, &errstr) < 0) {
+        flux_log_error (h, "content-s3 create bucket");
+        goto error;
+    }
+
+    ctx->cfg = cfg;
+    
+    if (flux_msg_handler_addvec (h, htab, ctx, &ctx->handlers) < 0)
+        goto error;
+
+    return ctx;
+
+error:
+    content_s3_destroy(ctx);
+    return ctx;
+}
+
+int mod_main (flux_t *h, int argc, char **argv)
+{
+    struct content_s3 *ctx;
+
+    if (!(ctx = content_s3_create (h))) {
+        flux_log_error (h, "content_s3_create failed");
+        return -1;
+    }
+    if (content_register_backing_store (h, "content-s3") < 0)
+        goto done;
+    if (content_register_service (h, "content-backing") < 0)
+        goto done;
+    if (content_register_service (h, "kvs-checkpoint") < 0)
+        goto done;
+    if (flux_reactor_run (flux_get_reactor (h), 0) < 0) {
+        flux_log_error (h, "flux_reactor_run");
+        goto done;
+    }
+    if (content_unregister_backing_store (h) < 0)
+        goto done;
+done:
```

Comment:
You inherited this bug from me (sorry).  All the `goto done` calls should lead to a return -1 (with errno set) so that `flux module load` fails if any of these things fail.

One way to do it is create an `int rc = -1` and replace the `return 0` with `return rc`, then set `rc = 0` right before the `done:` label.

Extra credit to fix in `content-files` too (in a standalone commit with a `Problem:` style commit message - see `git log` for examples).